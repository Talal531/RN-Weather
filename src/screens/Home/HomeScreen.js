import React, {useState, useEffect, useCallback} from 'react';
import {
  Text,
  View,
  ImageBackground,
  Image,
  TouchableOpacity,
  SafeAreaView,
  Platform,
} from 'react-native';
import {useNavigation} from 'react-navigation-hooks';
import {requestAndroidPermission} from '../../services/permissionService';
import Geolocation from '@react-native-community/geolocation';
import {_getWeatherByLocation} from '../../services/weatherServices';
import {useSelector, useDispatch} from 'react-redux';
import {storeWeather, updatelocation} from '../../store/actions';
import {HomeStyles} from './HomeScreenStyles';

//functional components with hooks
export default function HomeScreen() {
  const {navigate} = useNavigation();
  const [loading, setLoading] = useState(false);
  const weather = useSelector(state => state.weather);
  const location = useSelector(state => state.location);
  const dispatch = useDispatch();

  useEffect(() => {
    async function checkForPermissions() {
      // You can await here
      let isGranted = await requestAndroidPermission();
      console.log('permission', isGranted);

      if (isGranted || Platform.OS == 'ios') {
        getCurrentLocation();
      }
    }
    checkForPermissions();
  }, []);

  useEffect(() => {
    async function getWeatherDetails() {
      let weather = await _getWeatherByLocation(location);

      setLoading(false);
      if (weather) getWeather(weather);
    }
    getWeatherDetails();
  }, [location]);

  function getCurrentLocation() {
    setLoading(true);
    Geolocation.getCurrentPosition(
      location => {
        getLocation(location);
      },
      error => {
        console.log('err', error);
        setLoading(false);
      },
    );
  }

  const getWeather = useCallback(data => dispatch(storeWeather(data)), [
    dispatch,
  ]);

  const getLocation = useCallback(data => dispatch(updatelocation(data)), [
    dispatch,
  ]);

  function renderLocation() {
    console.log('renderLocation', weather);
    if (loading) {
      return <Text style={HomeStyles.heading}>Loading</Text>;
    } else if (!location || !weather) {
      return (
        <Text style={HomeStyles.heading}>Couldn't get current location</Text>
      );
    } else {
      return (
        <TouchableOpacity
          onPress={() => {
            navigate('Detail');
          }}>
          <Text style={HomeStyles.heading}>{` ${weather.name},`}</Text>
          <Text style={HomeStyles.heading}>{` ${weather.sys.country}`}</Text>
        </TouchableOpacity>
      );
    }
  }

  return (
    <ImageBackground
      source={require('assets/imgs/bg.png')}
      resizeMode="cover"
      style={HomeStyles.container}>
      <View style={HomeStyles.header}>
        <TouchableOpacity onPress={() => {}}>
          <Image
            style={HomeStyles.closeImg}
            source={require('assets/imgs/close.png')}
          />
        </TouchableOpacity>

        <View style={HomeStyles.liveButton}>
          <Text style={HomeStyles.liveButtonText}>Live</Text>
        </View>
      </View>

      <View style={HomeStyles.content}>
        <TouchableOpacity
          onPress={() => getCurrentLocation()}
          style={HomeStyles.currentLocationBox}>
          <Image
            style={HomeStyles.currentLocationBoxImg}
            source={require('assets/imgs/mark.png')}
          />
          <Text style={HomeStyles.currentLocationBoxText}>
            CURRENT LOCATION
          </Text>
        </TouchableOpacity>

        {renderLocation()}
      </View>
    </ImageBackground>
  );
}
