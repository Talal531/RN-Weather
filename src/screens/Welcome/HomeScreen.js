import React, {Component} from 'react';
import {View, Text} from 'react-native';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

export class HomeScreen extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }
  render() {
    return (
      <View>
        <Text> prop </Text>
      </View>
    );
  }
}

const mapStateToProps = state => ({});

const mapDispatchToProps = {};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(HomeScreen);
