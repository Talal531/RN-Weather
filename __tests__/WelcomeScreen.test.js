import React from 'react';

import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store';

import renderer from 'react-test-renderer';

import HomeScreen from '../src/screens/Home/HomeScreen';

const middlewares = [];
const mockStore = configureStore(middlewares);

const INITIAL_STATE = {};

const storeWeather = () => ({
  type: 'STORE_WEATHER',
  payload: INITIAL_STATE,
});

const storeWeatherHourly = () => ({
  type: 'STORE_WEATHER_HOURLY',
  payload: INITIAL_STATE,
});

describe('Home Screen', () => {
  const store = mockStore(INITIAL_STATE);

  it('should renders correctly', async () => {
    const tree = renderer.create(
      <Provider store={store}>
        <HomeScreen />
      </Provider>,
    );
    const json = tree.toJSON();
    expect(json).toMatchSnapshot();
  });

  it('getweather details', async () => {
    const tree = renderer.create(
      <Provider store={store}>
        <HomeScreen />
      </Provider>,
    );
    const json = tree.getInstance();
  });

  it('storeWeatherHourly', () => {
    store.dispatch(storeWeatherHourly());

    const actions = store.getActions();

    const expectedPayload = {
      type: 'STORE_WEATHER_HOURLY',
      payload: INITIAL_STATE,
    };
    expect(actions).toEqual([expectedPayload]);
  });
});
