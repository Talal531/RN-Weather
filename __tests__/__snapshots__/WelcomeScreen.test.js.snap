// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen should renders correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "flex": 1,
      "paddingHorizontal": 23.57142857142857,
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "testUri": "../../../assets/imgs/bg.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 172.85714285714283,
        "paddingVertical": 23.57142857142857,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/imgs/close.png",
          }
        }
        style={
          Object {
            "height": 26.714285714285715,
            "width": 26.714285714285715,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#ff2b57",
          "borderRadius": 11,
          "paddingHorizontal": 23.57142857142857,
          "paddingVertical": 7.857142857142857,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 18.857142857142858,
          }
        }
      >
        Live
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "paddingHorizontal": 15.714285714285714,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "flex-start",
          "backgroundColor": "rgba(0,0,0,0.1)",
          "borderRadius": 11,
          "flexDirection": "row",
          "marginBottom": 62.857142857142854,
          "opacity": 1,
          "paddingHorizontal": 23.57142857142857,
          "paddingVertical": 11,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../assets/imgs/mark.png",
          }
        }
        style={
          Object {
            "height": 23.57142857142857,
            "resizeMode": "contain",
            "width": 31.428571428571427,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "#ffffff",
            "fontSize": 18.857142857142858,
            "marginLeft": 7.857142857142857,
          }
        }
      >
        CURRENT LOCATION
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "#ffffff",
          "fontSize": 55,
        }
      }
    >
      Couldn't get current location
    </Text>
  </View>
</View>
`;
